README.md
# trylip
The aim of this project is to detect various types of expressions such as Smiling, Laughing, Pouting, Opening mouth etc using human lip movement in real time.
The main challenge of detecting any kind of lip movement is to find some landmark points on the lips. To do this, I first 
localized the face in the frame and then used the facial landmark detector of the dlib library -- an implementation of this 
work -- to detect the lip landmarks on the face. From these points, I crafted some features and then used k-means clustering 
algorithm to partition between five types of lip states. The clusters being learned, I manually assigned some labels to them and thus completing this project. The end result is a lip movement detector that can detect five types of lip movements in real-time.
The detector is robust to varying distances from the camera. Here is a demonstration of this project.

![No Movement](No_movement.png?raw=true "No Movement") 
![Smiling](Smiling.png?raw=true "Smiling")
![Laughing](Lauging.png?raw=true "Laughing")
![Pouting](Pouting.png?raw=true "Pouting")
![Opening](Open.png?raw=true "Opening")

# Face Localization:
To localize face in a frame, I used dlibs frontal face detector.


# Detecting Landmarks:
To detect the landmarks on the lips, I used dlibs landmark detection library. There are 20 such points (from index 49 to 68). The indices of the points are given in the following figure.

![Landmark](Landmarks.png?raw=true "Visualizing each of the 68 facial coordinate points from the iBUG 300-W dataset ")

# Extracting Features From Landmarks:
To extract features from landmarks, I have done the following two steps:
Take the points with index 49,  51, 52, 53, 55, 57, 59.
Create a polygon using these points and take the interior angles of the polygon as feature.

# Unsupervised Learning model:
Now that I have some features to work with, I used k-means clustering to differentiate between the shapes of the polygon. To train the model, I just recorded a video of me demonstrating different shapes. The model was able to learn the clusters just by watching one minute of video which is fascinating.

# Future Work:
I intend to extend this work to detect various shapes of mouth when human talks. My aim is to interpret what a person is saying just by analyzing the lip movement. This can help mutes to talk about using vocal chords.


history
(lp0
(lp1
cnumpy.core.multiarray
scalar
p2
(cnumpy
dtype
p3
(S'f8'
p4
I0
I1
tp5
Rp6
(I3
S'<'
p7
NNNI-1
I-1
I0
tp8
bS'\x07\xd5T{\xee\xb9#@'
p9
tp10
Rp11
ag2
(g6
S'\x00\x00\x00\x00\x00\x00\xe0?'
p12
tp13
Rp14
ag2
(g6
S'\xb6}\xfcI\x87\xfa\x15@'
p15
tp16
Rp17
ag2
(g6
S'\x12NC\xfd\xf0\x1d\xd6?'
p18
tp19
Rp20
ag2
(g6
S'\x04\xb2\x0eu\xb4\xb8\xf9?'
p21
tp22
Rp23
ag2
(g6
S'd\xfd\xd5\xeb\xe1,\xcb?'
p24
tp25
Rp26
ag2
(g6
S'\xb1\xdby\xb2\x19\xe2\xe0?'
p27
tp28
Rp29
ag2
(g6
S'\x8a\xe6\xfb\x85\xaa\xe5\xb3?'
p30
tp31
Rp32
ag2
(g6
S"8\x01('s\xbe\x0e@"
p33
tp34
Rp35
ag2
(g6
S'q\xcc;$\xc0\x82\xc3?'
p36
tp37
Rp38
ag2
(g6
S'h2\xb7(\x0f\xb2\x1b@'
p39
tp40
Rp41
ag2
(g6
S'\xb0\xc9\xddKn\x82\xd4?'
p42
tp43
Rp44
ag2
(g6
S'\xb9\xf46W\xc8""@'
p45
tp46
Rp47
ag2
(g6
S'\x00\x00\x00\x00\x00\x00\xe0?'
p48
tp49
Rp50
ag2
(g6
S'B)\xa8\xe5\xfdG\x0f@'
p51
tp52
Rp53
ag2
(g6
S'\x18a\x82}\xd4\xc0\xd6?'
p54
tp55
Rp56
ag2
(g6
S'Gz\xf8i\xe2\\\xdb?'
p57
tp58
Rp59
ag2
(g6
S'\xb5\xe0\x92\xd3\x997\xcf?'
p60
tp61
Rp62
ag2
(g6
S'\xea\xd3\xbe\x89|4\x10@'
p63
tp64
Rp65
ag2
(g6
S'\x8e\xa4b\xdb\xa4\x97\xcb?'
p66
tp67
Rp68
ag2
(g6
S'\n]e\xa7\xe0f\x1a@'
p69
tp70
Rp71
ag2
(g6
S'\x16\x00\xcd]#\xd7\xd0?'
p72
tp73
Rp74
ag2
(g6
S'8s\x14S\xa8\xa4!@'
p75
tp76
Rp77
ag2
(g6
S'\xaf\xebo\x8f\x1f\xf6\xd6?'
p78
tp79
Rp80
ag2
(g6
S'\xf7\\\x80\xe0\x01\x9b @'
p81
tp82
Rp83
ag2
(g6
S'AAGa\x98\x89\xda?'
p84
tp85
Rp86
ag2
(g6
S'\xb9\xb3\xaa6\xad\xd9\r@'
p87
tp88
Rp89
ag2
(g6
S'\x18a\x82}\xd4\xc0\xc6?'
p90
tp91
Rp92
ag2
(g6
S'&\xe4\x0e{\xd0\xe2\xf2?'
p93
tp94
Rp95
ag2
(g6
S'\xe5\x0c\xf2L\xea\xad\xa8?'
p96
tp97
Rp98
ag2
(g6
S'\x8e\xbcC\xec\x10\xf1\x00@'
p99
tp100
Rp101
ag2
(g6
S'"\xc9\xd4\xb6/\x8e\xc0?'
p102
tp103
Rp104
ag2
(g6
S'(!\xa3\xb1je\x1d@'
p105
tp106
Rp107
ag2
(g6
S'\x10\x83\x82S\x89[\xda?'
p108
tp109
Rp110
ag2
(g6
S'gM\x93\xb3h<\x04@'
p111
tp112
Rp113
ag2
(g6
S'6\xc7y\x9a\xa5\xc8\xc0?'
p114
tp115
Rp116
ag2
(g6
S'\xf3d\xc2z\x83\xa3\xe8?'
p117
tp118
Rp119
ag2
(g6
S'\xb5\xe0\x92\xd3\x997\x9f?'
p120
tp121
Rp122
ag2
(g6
S'dCt\xf0\x8c\x8e\n@'
p123
tp124
Rp125
ag2
(g6
S'\xf5\xaaJO\xe4>\xc6?'
p126
tp127
Rp128
aa(lp129
g2
(g6
S'\x97\x87\x83\xf7\xa2\xf1"@'
p130
tp131
Rp132
ag2
(g6
S'H\n}\xbd+M\xdf?'
p133
tp134
Rp135
ag2
(g6
S'\xbd\xb3<^\x0b\xc3\x12@'
p136
tp137
Rp138
ag2
(g6
S'VUUUUU\xd5?'
p139
tp140
Rp141
ag2
(g6
S'\xd8\xebdF\xea\xd4\xf2?'
p142
tp143
Rp144
ag2
(g6
S'\xb3[\x7f\xc2H\x07\xcd?'
p145
tp146
Rp147
ag2
(g6
S'\xff\xed@\xda|w\xee?'
p148
tp149
Rp150
ag2
(g6
S'\x81v\x87\x1e\xe37\xb9?'
p151
tp152
Rp153
ag2
(g6
S'\x02\xe6\xe1<4x\x12@'
p154
tp155
Rp156
ag2
(g6
S'\xef\xbf\xfe\xa3\x97\xec\xc7?'
p157
tp158
Rp159
ag2
(g6
S'\xf8\xda\x81\xa5\xcea\x1d@'
p160
tp161
Rp162
ag2
(g6
S'\x05\x05xg\xfdx\xd5?'
p163
tp164
Rp165
ag2
(g6
S'\x0f\xd1U%\xc1!"@'
p166
tp167
Rp168
ag2
(g6
S'\xb9\xb8\\\x05sY\xe0?'
p169
tp170
Rp171
ag2
(g6
S'\xf9\xf1\xf4:.9\x11@'
p172
tp173
Rp174
ag2
(g6
S'T\xcd,.\xf2\xf2\xd6?'
p175
tp176
Rp177
ag2
(g6
S'\xa6dj{\x9f\x12\xa4?'
p178
tp179
Rp180
ag2
(g6
S'\xf5P\xeb\xfc\xbb\xf8\xcd?'
p181
tp182
Rp183
ag2
(g6
S'\xe3\x10K@\rv\n@'
p184
tp185
Rp186
ag2
(g6
S'n\x0b\x00n\x16\x88\xc9?'
p187
tp188
Rp189
ag2
(g6
S'\xc0\xd3\x8d\xf6%\xda\x18@'
p190
tp191
Rp192
ag2
(g6
S'x\x1d>;\xd6\x12\xd0?'
p193
tp194
Rp195
ag2
(g6
S'KPqJ\x00\xdc @'
p196
tp197
Rp198
ag2
(g6
S'\x03F\xc1\x10B\xbb\xd6?'
p199
tp200
Rp201
ag2
(g6
S'\xf1\x95\xd6\x84[\xac\x1f@'
p202
tp203
Rp204
ag2
(g6
S'6\xef \x89\xb7\xb9\xd9?'
p205
tp206
Rp207
ag2
(g6
S'Y0\nh\x12=\x07@'
p208
tp209
Rp210
ag2
(g6
S'K#\xac/R\x8e\xc6?'
p211
tp212
Rp213
ag2
(g6
S'\xef\xb7\x14h\x1e\xe7\xd6?'
p214
tp215
Rp216
ag2
(g6
S'K#\xac/R\x8e\x96?'
p217
tp218
Rp219
ag2
(g6
S'\xc57\xc7"\x08\x04\x07@'
p220
tp221
Rp222
ag2
(g6
S'\x9dJT\xb3\xd3\xef\xc3?'
p223
tp224
Rp225
ag2
(g6
S'\xe5\x9d\x83\x0b\x8f\t\x1f@'
p226
tp227
Rp228
ag2
(g6
S'\xf8a\xc0\x86\xf4\x1d\xdb?'
p229
tp230
Rp231
ag2
(g6
S'\xc57\xc7"\x08\x04\x07@'
p232
tp233
Rp234
ag2
(g6
S'\x9dJT\xb3\xd3\xef\xc3?'
p235
tp236
Rp237
ag2
(g6
S'\xb9\xaa44\xe5\xad\xe8?'
p238
tp239
Rp240
ag2
(g6
S'\x96\xaaS\x85\x1f\xe6\x9f?'
p241
tp242
Rp243
ag2
(g6
S'Y0\nh\x12=\x07@'
p244
tp245
Rp246
ag2
(g6
S'K#\xac/R\x8e\xc6?'
p247
tp248
Rp249
aa(lp250
g2
(g6
S'\xae3Q\x14\x85}$@'
p251
tp252
Rp253
ag2
(g6
S'\xa8\x01@\xd2Cq\xdf?'
p254
tp255
Rp256
ag2
(g6
S'\x19\x11\xef\xe4z\xd3\x15@'
p257
tp258
Rp259
ag2
(g6
S'\x8c \xd4@r\x97\xd5?'
p260
tp261
Rp262
ag2
(g6
S'\xc1:n\xd8\xb0A\xf3?'
p263
tp264
Rp265
ag2
(g6
S"\xe5'i\x88\xd6\x17\xc8?"
p266
tp267
Rp268
ag2
(g6
S'oK\x97\xac\x10n\xe1?'
p269
tp270
Rp271
ag2
(g6
S'\xbf\x8c1\x84\xabl\xb3?'
p272
tp273
Rp274
ag2
(g6
S'\xeb\xe4FJ\xa9\x1a\x11@'
p275
tp276
Rp277
ag2
(g6
S'/>\t@\xd5\t\xc5?'
p278
tp279
Rp280
ag2
(g6
S'\xf6q\xf1\x04\x8b\xa6\x1b@'
p281
tp282
Rp283
ag2
(g6
S'\xa3\xb8m\xfd\xb5\x05\xd4?'
p284
tp285
Rp286
ag2
(g6
S'\x93\x90\x81>\xf0\xb2"@'
p287
tp288
Rp289
ag2
(g6
S'\xe8\xc2:\xe6\xb5I\xe0?'
p290
tp291
Rp292
ag2
(g6
S'Q\x07\xd9<\xee\xe0\x0e@'
p293
tp294
Rp295
ag2
(g6
S'\xab\x88\xf16w6\xd6?'
p296
tp297
Rp298
ag2
(g6
S'\xb1\xf8\x03\xb3~\x84\xde?'
p299
tp300
Rp301
ag2
(g6
S'\x10 \xb9<\xc4y\xce?'
p302
tp303
Rp304
ag2
(g6
S'D\xe1{\xf3W\xe4\r@'
p305
tp306
Rp307
ag2
(g6
S'\xbbF)\xfc>\x12\xcb?'
p308
tp309
Rp310
ag2
(g6
S'\xc1\xabn\x06\xa5\x0c\x19@'
p311
tp312
Rp313
ag2
(g6
S'(\xa5\xd6{\xe0i\xcf?'
p314
tp315
Rp316
ag2
(g6
S'\x8ej\x03\x06u\xaa!@'
p317
tp318
Rp319
ag2
(g6
S'\t~:4~j\xd6?'
p320
tp321
Rp322
ag2
(g6
S'\xb6\xf8\x99\xb7\xf5\x8d @'
p323
tp324
Rp325
ag2
(g6
S'+\xc8q\xf67\xe8\xd9?'
p326
tp327
Rp328
ag2
(g6
S'\x99k\r\xa2\xc5t\n@'
p329
tp330
Rp331
ag2
(g6
S'\xe0\x03\xf92\x9d\xb7\xc5?'
p332
tp333
Rp334
ag2
(g6
S'\x1a\x8d\x0b\x96\xa0\xfc\xf2?'
p335
tp336
Rp337
ag2
(g6
S"\xe5'i\x88\xd6\x17\xa8?"
p338
tp339
Rp340
ag2
(g6
S'\xc13\x8f>n\xd6\x00@'
p341
tp342
Rp343
ag2
(g6
S'\xe7\xd5S\xf3\x82)\xc0?'
p344
tp345
Rp346
ag2
(g6
S'_n\xa2\xb8UB\x1d@'
p347
tp348
Rp349
ag2
(g6
S'\xda\xd8\r\xff@\xbb\xd9?'
p350
tp351
Rp352
ag2
(g6
S'\xc1\xd7vD\x8e\xed\x06@'
p353
tp354
Rp355
ag2
(g6
S'\n\xb4\xf3\xa5\x1a\x0c\xc3?'
p356
tp357
Rp358
ag2
(g6
S'[&\xac\x9c\x8e\xb4\xe8?'
p359
tp360
Rp361
ag2
(g6

tools.py
import cv2
import dlib
from imutils import face_utils
import imutils
import numpy as np
from numpy.linalg import norm

def bb_to_rect(bb):
    print bb
    x= dlib.rectangle(int(bb[0]), int(bb[1]), int(bb[0] + bb[2]), int(bb[0] + bb[3]))
    print x
    return x

def detectFace(img, option = 1):
    predictor = dlib.shape_predictor('shape_predictor_68_face_landmarks.dat')
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    faces = []
    if option == 1:
        detector = dlib.get_frontal_face_detector()
        rects = detector(gray, 1)
        for (i, rect) in enumerate(rects):
            shape = predictor(gray, rect)
            shape = face_utils.shape_to_np(shape)
            print type(shape)
            faces.append((face_utils.rect_to_bb(rect), shape))
    elif option == 2:
        detector = cv2.CascadeClassifier('/home/turja/opencv-3.1.0/data/haarcascades/haarcascade_frontalface_default.xml')
        face_roi = detector.detectMultiScale(gray, 1.3, 5)
        for face in face_roi:
            # determine the facial landmarks for the face region, then
            # convert the facial landmark (x, y)-coordinates to a NumPy
            # array
            rect = bb_to_rect(face)
            shape = predictor(gray, rect)
            shape = face_utils.shape_to_np(shape)
            faces.append((face, shape))

    return faces

def detectLipMovement(mouth, scale = 1):
    # take 3 points from the lips
    point_1 = np.array([mouth[6][0], mouth[6][1]])
    point_2 = np.array([mouth[3][0], mouth[3][1]])
    point_3 = np.array([mouth[9][0], mouth[9][1]])

    # create two vectors
    v1= point_2 - point_1
    v2 = point_3 - point_1

    #find angles between them
    c = np.dot(v1, v2)/norm(v1)/norm(v2)
    return np.arccos(np.clip(c, -1, 1))/scale

def detectLipMovement2(mouth, scale = 1):
    pivots = [mouth[13], mouth[14], mouth[15]]
    target = [mouth[19], mouth[18], mouth[17]]

    dist = 0
    for a, b in zip(pivots, target):
        dist = dist + norm(a - b)
    return dist/(len(mouth))/scale

def detectLipMovement3(mouth, scale = 1):
    list1 = np.concatenate((mouth[0:6], mouth[12:16]), axis=0)
    print len(list1)
    list2 = np.concatenate((mouth[7:12], mouth[16:]))
    print len(list2)
    dist = 0
    for m1 in list1:
        for m2 in list2:
            dist =  dist + norm(m1 - m2)

    return dist/len(list1)/len(list2)/scale

def find_angle(v1, v2):
    print('Vect1', v1)
    print('Vect2', v2)
    print v2
    c = np.dot(v1, v2) / norm(v1) / norm(v2)
    angle = np.arccos(np.clip(c, -1, 1))
    if angle <= 180:
        return angle
    else:
        return 360 - angle

def makeLine(p1, p2):
    return np.array(p1-p2)
    #return np.array(p2[0], p2[1]) - np.array(p1[0], p1[1])

def find_mid(p, q):
    return ((p[0] + q[0])/2, (p[1] + q[1])/2)
def featureSet(points, scale):
    pivot = find_mid(points[0], points[6])

    angle_feature_set = []
    for p in points:
        angle_feature_set.append(find_angle_from_points(points[6], pivot, p))
        angle_feature_set.append(norm(p - pivot)/scale)

    return angle_feature_set

def find_angle_from_points(p1, p2, p3):
    print 'Finding angle from 3 points'
    v1 = makeLine(p2, p1)
    v2 = makeLine(p2, p3)
    return find_angle(v1, v2)

def featureSetUsingHull(points, scale):
    hull = cv2.convexHull(points)
    feat = []
    hull_len = len(hull)
    #adding distance in feature set
    for i in range(0, len(hull)):
        p1 = hull[i]
        p2 = hull[(i + 1)%hull_len]
        feat.append(norm(p1[0] - p2[0])/scale)


    #adding angles in feature set
    for i in range(0, len(hull)):
        p1 = hull[i]
        p2 = hull[(i + 1) % hull_len]
        p3 = hull[(i + 2) % hull_len]

        print p1[0], p2[0], p3[0]
        feat.append(find_angle_from_points(p1[0], p2[0], p3[0]))

    print 'Lenn: :',len(feat)
    return feat

def featureSetUsingOnlyImprotantPoints(points):
    pIndex = [0, 2, 3, 4, 6, 8, 10]
    feat = []

    for i in range(0, len(pIndex)):
        feat.append(find_angle_from_points(points[pIndex[i]], points[pIndex[(i + 1)%len(pIndex)]], points[pIndex[(i + 2)%len(pIndex)]]))

    feat.append(norm(points[0] - points[6]) * 1.0 / norm(points[3] - points[9]))
    return feat

def find_nearest_neighbor(val, seedList):
    dist = -1
    nseed = None
    count = -1
    index = 0
    for seed in seedList:
        count = count + 1
        p = norm(np.array(val) - np.array(seed))
        print p
        if dist == -1 or p < dist:
            dist = p
            nseed = seed
            index = count

    return nseed, index





seedsGen.py
import os
import tools
import cv2
from numpy.linalg import norm
import pickle
import glob
from matplotlib import pyplot as plt

frame = cv2.imread('Seeds/seed1.jpg')
print frame
path = 'Seeds/'
fl = open('seed', 'w+')
seedfeat = []

for i in range(1,6):
    pathName = path + 'seed' + str(i) + '.jpg'
    print pathName
    frame = cv2.imread(pathName)
    roi_faces = tools.detectFace(frame)
    for face, landmarks in roi_faces:
        print 'I am here'
        x, y, w, h = face
        i = 48
        j = 68
        val = tools.detectLipMovement2(landmarks[i: j], scale=norm(landmarks[0] - landmarks[16]))
        afs = tools.featureSet(landmarks[i: j], scale=norm(landmarks[0] - landmarks[16]))
        print len(afs)
        seedfeat.append(afs)




pickle.dump(seedfeat, fl)
fl.close()
print seedfeat
with open('seed', 'rb') as fl:
    print pickle.load(fl)

main.py
import numpy as np
import tools
import cv2
from matplotlib import pyplot as plt
import time
import datetime
from numpy.linalg import norm
import sounddevice as sd
from sklearn.cluster import KMeans
from sklearn.externals import joblib
import pickle


cap = cv2.VideoCapture(0)
graph_data_set = []
stage = 0
'''
#setting up sound recording environment
sd.default.samplerate = 44100
fs = 44100
duration = 10
sound = sd.rec(int(duration * fs), samplerate=fs, channels=1)'''

featureSet = []
kmeans = joblib.load('kmeansModel.pkl')
seeds = pickle.load(open('seed', 'rb'))
frameLabel = 0
featureSet = []
while(cap.isOpened()):
    # Capture frame-by-frame
    ret, frame = cap.read()
    frameLabel = frameLabel + 1

    if ret == True:
        fourcc = cv2.VideoWriter_fourcc(*'XVID')
        out = cv2.VideoWriter('output.avi', fourcc, 30, (640,480))
        # Our operations on the frame come here
        #gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
        roi_faces = tools.detectFace(frame, 1)
        for face, landmarks in roi_faces:
            x, y, w, h = face
            i = 48
            j = 68
            cv2.rectangle(frame, (x, y), (x + w, y + h), (255, 0, 0), 2)
            # loop over the subset of facial landmarks, drawing the
            # specific face part
            '''
            (p, q) = landmarks[48]
            (s, t) = landmarks[54]
            (p, q) = ((p + s)/2, (q + t)/2)
            for (x, y) in landmarks[i:j]:
                cv2.circle(frame, (x, y), 1, (0, 0, 255), -1)
                cv2.line(frame, (x, y), (p, q), (0, 255, 0), 1)'''

            npl = landmarks[i:j]
            pIndex = [0, 2, 3, 4, 6, 8, 10]

            for a in range(0, len(pIndex)):
                (p1, q1) = npl[pIndex[a]]
                (p2, q2) = npl[pIndex[(a+1)%len(pIndex)]]
                cv2.circle(frame, (p1, q1) , 1, (0, 0, 255), -1)
                cv2.line(frame, (p1, q1) , (p2, q2), (0, 255, 0), 1)

            ts = time.time()
            st = datetime.datetime.fromtimestamp(ts).strftime('%Y-%m-%d %H:%M:%S')
            #val = tools.detectLipMovement2(landmarks[i : j], scale = norm(landmarks[0] - landmarks[16]))
            afs = tools.featureSetUsingOnlyImprotantPoints(landmarks[i : j])
            featureSet.append(afs)
            temp = []
            temp.append(afs)
            featureSet.append(afs)
            temp = np.array(temp)
            where_are_NaNs = np.isnan(temp)
            temp[where_are_NaNs] = 0
            #val1 = 3
            val1 = kmeans.predict(np.array(temp))
            #minseed, val1 = tools.find_nearest_neighbor(afs, seeds)
            print('kmean prediction: ', val1)
            #graph_data_set.append((val, st))
            writeText = None
            if val1 == 2:
                writeText = 'Smiling'
            elif val1 == 1:
                writeText = 'Laughing'
            elif val1 == 0:
                writeText = 'Pouting'
            elif val1 == 3:
                writeText = 'No movement'
            elif val1 == 4:
                writeText = 'Open'

            cv2.putText(frame, writeText, (x + 40, y - 10),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)


        # Display the resulting frame
        cv2.imshow('frame',frame)
        cv2.imwrite('output.jpg', frame)
        out.write(frame)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break
    else:
        break

# When everything done, release the capture
cap.release()
out.release()
cv2.destroyAllWindows()
i = 0
'''smooth_data = []
graph_data, time = zip(*graph_data_set)
for i in range(1, len(graph_data) - 1):
    smooth_data.append((graph_data[i - 1] + 3*graph_data[i] + graph_data[i + 1])/5)'''

'''smooth_dif = []
th = 0.7
for i in range(1, len(diff) - 1):
    smooth_dif.append((diff[i - 1] + 3 * diff[i] + diff[i + 1])/5)
    if i > 0 and smooth_dif[i] - smooth_dif[i - 1] > th:'''

# cancel noise in sound
'''for i in range(2, len(sound)-2):
    a = 0
    for j in range(-2,3):
        a = a + sound[i + j]
    sound[i] = a/5'''

#show response graph
#plt.plot(graph_data)
'''
plt.subplot(2,1,1)
plt.plot(sound)
plt.title('Actual Sound')
plt.subplot(2,1,2)
plt.plot(smooth_data)
plt.title('Estimation from lips')
plt.show()

plt.plot(graph_data)
plt.show()
'''
#Kmeans
#print featu.shape
'''
print len(featureSet)
with open('history_three', 'a+b') as f:
    pickle.dump(featureSet, f)



with open('history_three', 'rb') as f:
    while 1:
        try:
            tmpfeat = pickle.load(f)
            for elem in tmpfeat:
                featureSet.append(elem)
        except EOFError:
            break

print type(featureSet)
featureSet = np.array(featureSet)
where_are_NaNs = np.isnan(featureSet)
featureSet[where_are_NaNs] = 0
#seeds = pickle.load(open('seed', 'rb'))
#seeds = np.array(seeds)
#print seeds.shape
kmeans = KMeans(n_clusters=5, random_state=0).fit(featureSet)
#joblib.dump(kmeans, 'kmeansModel.pkl')'''


